src/catalog/src/transaction.rs

```diff
@@ -168,29 +168,90 @@ pub struct Transaction<'a> {
 impl<'a> Transaction<'a> {
     pub fn new(
         durable_catalog: &'a mut dyn DurableCatalogState,
-        databases: BTreeMap<proto::DatabaseKey, proto::DatabaseValue>,
-        schemas: BTreeMap<proto::SchemaKey, proto::SchemaValue>,
-        roles: BTreeMap<proto::RoleKey, proto::RoleValue>,
-        items: BTreeMap<proto::ItemKey, proto::ItemValue>,
-        comments: BTreeMap<proto::CommentKey, proto::CommentValue>,
-        clusters: BTreeMap<proto::ClusterKey, proto::ClusterValue>,
-        cluster_replicas: BTreeMap<proto::ClusterReplicaKey, proto::ClusterReplicaValue>,
-        introspection_sources: BTreeMap<
-            proto::ClusterIntrospectionSourceIndexKey,
-            proto::ClusterIntrospectionSourceIndexValue,
-        >,
-        id_allocator: BTreeMap<proto::IdAllocKey, proto::IdAllocValue>,
-        configs: BTreeMap<proto::ConfigKey, proto::ConfigValue>,
-        settings: BTreeMap<proto::SettingKey, proto::SettingValue>,
-        timestamps: BTreeMap<proto::TimestampKey, proto::TimestampValue>,
-        system_gid_mapping: BTreeMap<proto::GidMappingKey, proto::GidMappingValue>,
-        system_configurations: BTreeMap<
-            proto::ServerConfigurationKey,
-            proto::ServerConfigurationValue,
-        >,
-        default_privileges: BTreeMap<proto::DefaultPrivilegesKey, proto::DefaultPrivilegesValue>,
-        system_privileges: BTreeMap<proto::SystemPrivilegesKey, proto::SystemPrivilegesValue>,
+        Snapshot {
+            databases,
+            schemas,
+            roles,
+            items,
+            comments,
+            clusters,
+            cluster_replicas,
+            introspection_sources,
+            id_allocator,
+            configs,
+            settings,
+            timestamps,
+            system_object_mappings,
+            system_configurations,
+            default_privileges,
+            system_privileges,
+        }: Snapshot,
     ) -> Result<Transaction, Error> {
+        let databases = databases
+            .into_iter()
```

Comment:
This looks like each Transaction will go through an additional protobuf decoding then encoding step? Is that right? Is the refactor worth the perf loss?

Reply:
Good point, what if we switch the types to use proto directly so we don't have to eat that cost?